### PR TITLE
HHH-18439 fix handling of null values in query cache

### PR DESCRIPTION
**Pull request by boskoop**
[HHH-18439](https://hibernate.atlassian.net/browse/HHH-18439): NullPointerException when access data from query cache

The problem was in JdbcValueCacheHit#getCurrentRowValue where a single null value in a cached query result would result in a NullPointerException.

----------------------
By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt)
and can be relicensed under the terms of the [LGPL v2.1 license](https://www.gnu.org/licenses/old-licenses/lgpl-2.1.txt) in the future at the maintainers' discretion.
For more information on licensing, please check [here](https://github.com/hibernate/hibernate-orm/blob/main/CONTRIBUTING.md#legal).

----------------------

[HHH-18439]: https://hibernate.atlassian.net/browse/HHH-18439?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: hibernate-core/src/test/resources/org/hibernate/orm/test/querycache/Item.hbm.xml [hunk 1]
```diff
@@ -17,7 +17,7 @@
 			<generator class="native"/>
 		</id>
 		<property name="name" not-null="true"/>
-		<property name="description" not-null="true"/>
```
Comment (boskoop) [inline hibernate-core/src/test/resources/org/hibernate/orm/test/querycache/Item.hbm.xml:20]: I had to make description nullable in the test entity in ordert to test the null mapping. To my understanding, this should not affect the existing tests.

File: hibernate-core/src/main/java/org/hibernate/sql/results/jdbc/internal/JdbcValuesCacheHit.java [hunk 2]
```diff
@@ -173,7 +173,10 @@ public Object getCurrentRowValue(int valueIndex) {
 			return null;
 		}
 		final Object row = cachedResults.get( position + offset );
-		if ( valueIndexesToCacheIndexes == null ) {
+		if ( row == null ) {
+			return null;
+		}
+		else if ( valueIndexesToCacheIndexes == null ) {
```
Comment (mbladel) [inline hibernate-core/src/main/java/org/hibernate/sql/results/jdbc/internal/JdbcValuesCacheHit.java:179]: I'm guessing the NPE was triggered by `row.getClass() != Object[].class`. Your change works, but I think we can use `instanceof` to be more concise:

```suggestion
		if ( valueIndexesToCacheIndexes == null || valueIndexesToCacheIndexes.length == 0 ) {
			return ( (Object[]) row )[valueIndex];
		}
		else if ( row instanceof Object[] ) {
			return ( (Object[]) row )[valueIndexesToCacheIndexes[valueIndex]];
		}
		else {
			assert valueIndexesToCacheIndexes[valueIndex] == 0;
			return row;
		}
```

Also, from looking at the code `valueIndexesToCacheIndexes` will never be null in the standard `JdbcValuesMapping` implementation - it will instead be an empty array, so it might be worth checking that as well.

File: hibernate-core/src/test/resources/org/hibernate/orm/test/querycache/Item.hbm.xml (hunk 1 above)
Comment (mbladel) [reply hibernate-core/src/test/resources/org/hibernate/orm/test/querycache/Item.hbm.xml:20]: This is fine. Could you please also add a test for a very simple `select null` query (like the one the user reported in the Jira) to assure that works as well?

**Comment by mbladel** [comment]
@boskoop I was curious, so I wrote a separate test with more assertions and also the `select null` case. I found this caused another problem when checking the return type of the query, so I have proposed fixes for both that and the NPE reported in the Jira in [this pr](https://github.com/hibernate/hibernate-orm/pull/8815).

**Comment by boskoop** [comment]
> @boskoop I was curious, so I wrote a separate test with more assertions and also the `select null` case. I found this caused another problem when checking the return type of the query, so I have proposed fixes for both that and the NPE reported in the Jira in [this pr](https://github.com/hibernate/hibernate-orm/pull/8815).

I think your solution is more elegant, also the tests are better since you are testing the session statistics for cache hits. I will close this pr in favor of your solution.
